interfaces/cython/cantera/composite.py

```diff
@@ -903,10 +923,24 @@ def join(species):
             self._output_dummy = self._indices
             self._shape = (rows,)
 
-        # restore data
-        for i in self._indices:
-            setattr(self._phase, mode, [st[i, ...] for st in state_data])
-            self._states[i] = self._phase.state
+        # use non-normalized state setters for Three property setter only
+        if mode in {"TPX", "TPY", "HPX", "HPY", "SPX", "SPY",
+                    "TDX", "TDY", "TPX", "TPY", "UVX", "UVY",
+                    "DPX", "DPY","HPX", "HPY", "SPX", "SPY",
+                    "SVX", "SVY"}:
```

Comment:
This change seems like the default will be to do non-normalized restoring of states, is that correct? Assuming that is correct, I'm not sure I agree that should be the default, I'd be interested to hear why you're making this change.

I guess the expectation for users is that the data they provide is exactly used to set the state, which is fair enough. I wonder whether another keyword argument should be used to control this behavior?

Last, I don't think it's necessary to provide an exhaustive list of the three-property setters, it should be sufficient to check that `len(mode) == 3`, right?

Reply:
From the documentation, the method `restore_data` is called internally to restored data saved to different file types( "csv", "h5" etc). We want the same exact state stored to be restored. That is why I think non-normalized should be the default.
There are some other three-property setters such as `TPQ`, `TDQ` that are not applicable in this context and that is what I am trying to catch. I agree with you that providing ab exhaustive list of the three-property setters might not be the best approach. Doing this might be better
``` if len(mode) == 3 and "Q" not in mode
```

Reply:
Yeah, I think it's a fair expectation on the part of the user to have exactly the input data restored. I think my nervous feeling is because this would be the only case that I can think of where a user action triggers non-normalized mole fraction setting by default. So in that sense it may be doing something a different user doesn't expect.

Whichever becomes the default here, it should definitely be possible to "restore" and have Cantera normalize mole fractions, for instance, if a set of initial conditions for a set of simulations is read from CSV, and the user specifies mole numbers instead of fractions.

Reply:
The `restore_data` method is called internally on previously saved `SolutionArray` object from Cantera and not on some user-defined set of initial conditions or a random "csv` or `h5` file format. We want the `SolutionArray` object data to be reloaded in the same format it was saved initially that is why I thought making non-normalized form the default would be applicable here. It gives the user a lot of flexibility as the user can now decide to go with the non-normalized or normalized form depending on the application. i.e, the `restore_data` is used to restore previously saved simulation results and we want the result to be consistent with the initial result saved.

Reply:
The signature of `restore_data` is currently
```Python
    def restore_data(self, data):
```
but it is also true that it is an internal function that is typically called by `read_csv`, `read_hdf` and `from_pandas`.

I believe it would make sense to add a keyword argument to `restore_data` with the default set to `normalize=False` (for reasons pointed out by @DavidAkinpelu). At least for `read_csv` and `from_pandas`, I believe that adding a similar keyword would allow a user to override the default behavior for cases as pointed out by @bryanwweber.